File: 641-dexign-circular-deque/MyCircularDeque.java
import java.util.*;

class MyCircularDeque {

    LinkedList<Integer> deque;
    int size = 0;

    public MyCircularDeque(int k) {
        deque = new LinkedList<>();
        size = k;
    }
    
    public boolean insertFront(int value) {
        if (isFull()) return false;

        deque.addFirst(value);
        return true;
    }
    
    public boolean insertLast(int value) {
        if (isFull()) return false;

        deque.addLast(value);
        return true;
    }
    
    public boolean deleteFront() {
        if(deque.isEmpty()) return false;
        
        deque.removeFirst();
        return true;
    }
    
    public boolean deleteLast() {
        if(deque.isEmpty()) return false;
        
        deque.removeLast();
        return true;
    }
    
    public int getFront() {
        if(deque.isEmpty()) return -1;
        
        return deque.getFirst();
    }
    
    public int getRear() {
        if(deque.isEmpty()) return -1;
        
        return deque.getLast();
    }
    
    public boolean isEmpty() {
        return deque.isEmpty();
    }
    
    public boolean isFull() {
        return deque.size() == size;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque obj = new MyCircularDeque(k);
 * boolean param_1 = obj.insertFront(value);
 * boolean param_2 = obj.insertLast(value);
 * boolean param_3 = obj.deleteFront();
 * boolean param_4 = obj.deleteLast();
 * int param_5 = obj.getFront();
 * int param_6 = obj.getRear();
 * boolean param_7 = obj.isEmpty();
 * boolean param_8 = obj.isFull();
 */

File: 641-dexign-circular-deque/MyCircularDeque.py
from collections import deque


class MyCircularDeque:

    def __init__(self, k: int):
        self.queue = deque()
        self.k = k     

    def insertFront(self, value: int) -> bool:
        if len(self.queue) == self.k:
            return False
        else:
            self.queue.appendleft(value)
            return True
        

    def insertLast(self, value: int) -> bool:
        if len(self.queue) == self.k:
            return False
        else:
            self.queue.append(value)
            return True
        

    def deleteFront(self) -> bool:
        if self.queue:
            self.queue.popleft()
            return True
        else:
            return False
        

    def deleteLast(self) -> bool:
        if self.queue:
            self.queue.pop()
            return True
        else:
            return False
        

    def getFront(self) -> int:
        if self.queue:
            return self.queue[0]
        else:
            return -1

    def getRear(self) -> int:
        if self.queue:
            return self.queue[len(self.queue)-1]
        else:
            return -1

    def isEmpty(self) -> bool:
        return len(self.queue) == 0

    def isFull(self) -> bool:
        return len(self.queue) == self.k